src/TorchSharp/DataLoader.cs

```diff
@@ -17,276 +17,352 @@ public static partial class utils
+        public class DataLoader : DataLoader<Dictionary<string, torch.Tensor>, Dictionary<string, torch.Tensor>>
+        {
+            /// <summary>
+            /// Pytorch style dataloader
+            /// </summary>
+            /// <param name="dataset">Dataset for create batch</param>
+            /// <param name="batchSize">Size of batch</param>
+            /// <param name="device">device for output tensor</param>
+            /// <param name="shuffler">Shuffler for dataloader</param>
+            /// <param name="num_worker">Count of worker</param>
+            /// <param name="drop_last">
+            /// Set to true to drop the last incomplete batch, if the dataset size is not divisible by the batch size.
+            /// If alse and the size of dataset is not divisible by the batch size, then the last batch will be smaller.
+            /// </param>
+            public DataLoader(Dataset dataset, int batchSize, IEnumerable<long> shuffler, Device device = null, int num_worker = 1, bool drop_last = false)
+                : base(dataset, batchSize, Collate, shuffler, device, num_worker, drop_last)
+            {
+            }
+
+            /// <summary>
+            /// Pytorch style dataloader
+            /// </summary>
+            /// <param name="dataset">Dataset for create batch</param>
+            /// <param name="batchSize">Size of batch</param>
+            /// <param name="shuffle">true if shuffle dataset, false for not</param>
+            /// <param name="device">device for output tensor</param>
+            /// <param name="seed">Seed for generating shuffle</param>
+            /// <param name="num_worker">Count of worker</param>
+            /// <param name="drop_last">
+            /// Set to true to drop the last incomplete batch, if the dataset size is not divisible by the batch size.
+            /// If alse and the size of dataset is not divisible by the batch size, then the last batch will be smaller.
+            /// </param>
+            public DataLoader(Dataset dataset, int batchSize, bool shuffle = false, Device device = null, int? seed = null, int num_worker = 1, bool drop_last = false)
+                : base(dataset, batchSize, Collate, shuffle, device, seed, num_worker, drop_last)
+            {
+            }
+
+            private static Dictionary<string, torch.Tensor> Collate(IEnumerable<Dictionary<string, torch.Tensor>> dic, torch.Device device)
+            {
+                Dictionary<string, torch.Tensor> batch = new();
+                foreach (var x in dic.First().Keys) {
+                    var t = cat(dic.Select(k => k[x].unsqueeze(0)).ToArray(), 0);
+                    if (t.device_type != device.type || t.device_index != device.index)
+                        t = t.to(device);
+                    batch[x] = t;
+                }
+                return batch;
+            }
+        }
+
+        public class IterableDataLoader : DataLoader<IList<torch.Tensor>, IList<torch.Tensor>>
```

Comment:
Do we generally try to have XML comments on types?

Reply:
We should. I'll add some.